x-pack/legacy/plugins/code/server/lib/esqueue/worker.js

```diff
@@ -465,7 +467,23 @@ export class Worker extends events.EventEmitter {
         if (jobs.length > 0) {
           this.debug(`${jobs.length} outstanding jobs returned`);
         }
-        return jobs;
+
+        // only return tasks for local repositories
+        const clusterState = this.clusterService.state();
+
+        const localRepos = clusterState.getNodeRepositories(this.clusterService.localNodeId);
+        if (!localRepos) {
+          return [];
+        }
+        const localRepoUrls = new Set(localRepos.map(repo => repo.url));
+        const localRepoUris = new Set(localRepos.map(repo => repo.uri));
+        // only need the jobs with `payload.payload.url` exists in the localRepoUrls
+        return jobs.filter(job => {
```

Comment:
There could be an issue in here in an ideal HA setup that all the top `this.checkSize` jobs returned may not belong to this node, while the actual queue jobs that belong to this node may fall beyond the `this.checkSize`. So that the `jobs.filter` will return 0 in here.

example: 
node `n1` pulls `j1, j2` jobs from ES, but neither of them belongs to node `n1` based on their payload repouri, while there is a job `j3` belongs to node `n1` still stays in ES.

To solves this, either creating a long ES filter contains all the repouris belong to the current node, or add a new `node` field when each queue job is created. Both have pros and cons. With current static routing table with only one data node, this issue won't be a problem. We may need to discuss this later. This might be another reason we need to switch to using local memory status instead of the current esqueue we are using now.

Reply:
Yes, this could be a problem, given that node may fail in the cluster-code mode, and the resources belong to the inactive node may become the blockers at the head of the queue. I  have some concerns on both modifications you mentioned, we can discuss this later.